api/client/joinservice.go

```diff
@@ -125,3 +131,71 @@ func (c *JoinServiceClient) RegisterUsingAzureMethod(ctx context.Context, challe
 	}
 	return certsResp.Certs, nil
 }
+
+// RegisterUsingTPMMethod registers the caller using the TPM join method and
+// returns signed certs to join the cluster. The caller must provide a
+// ChallengeResponseFunc which returns a *proto.RegisterUsingTPMMethodRequest
+// for a given challenge, or an error.
+func (c *JoinServiceClient) RegisterUsingTPMMethod(
+	ctx context.Context,
+	initReq *proto.RegisterUsingTPMMethodInitialRequest,
+	solveChallenge RegisterTPMChallengeResponseFunc,
+) (*proto.Certs, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tpmJoinClient, err := c.grpcClient.RegisterUsingTPMMethod(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = tpmJoinClient.Send(&proto.RegisterUsingTPMMethodRequest{
+		Payload: &proto.RegisterUsingTPMMethodRequest_Init{
+			Init: initReq,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "sending initial request")
+	}
+
+	challengeResp, err := tpmJoinClient.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err, "receiving challenge")
+	}
+
+	challenge, ok := challengeResp.Payload.(*proto.RegisterUsingTPMMethodResponse_ChallengeRequest)
+	if !ok {
+		return nil, trace.BadParameter(
+			"unexpected payload type %T, expected *RegisterUsingTPMMethodResponse_ChallengeRequest",
+			challengeResp.Payload,
+		)
+	}
+
+	solution, err := solveChallenge(challenge.ChallengeRequest)
+	if err != nil {
+		return nil, trace.Wrap(err, "calling solveChallenge")
+	}
+
+	err = tpmJoinClient.Send(&proto.RegisterUsingTPMMethodRequest{
+		Payload: &proto.RegisterUsingTPMMethodRequest_ChallengeResponse{
+			ChallengeResponse: solution,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "sending solution")
+	}
+
+	certsResp, err := tpmJoinClient.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err, "receiving certs")
+	}
+	certs, ok := certsResp.Payload.(*proto.RegisterUsingTPMMethodResponse_Certs)
+	if !ok {
+		return nil, trace.BadParameter(
+			"unexpected payload type %T, expected *RegisterUsingTPMMethodResponse_Certs",
+			certsResp.Payload,
+		)
+	}
+
+	return certs.Certs, nil
```

Comment:
Suggestion: similar to the one above. It's all equivalent, really.

```suggestion
	certs := certsResp.GetCerts()
	if certs == nil {
		return nil, trace.BadParameter(
			"unexpected payload type %T, expected *RegisterUsingTPMMethodResponse_Certs",
			certsResp.Payload,
		)
	}

	return certs, nil
```

Reply:
e999ea9928